### PR TITLE
feat(Harvest): Handle webauth connectors

### PR DESCRIPTION
**Pull request by doubleface**
Webauth is a protocol similar to OAuth but specific Budget Insight.

This PR needs https://github.com/cozy/cozy-stack/pull/2667 to be deployed in the stack to work properly + some specific secrets to be available.

For us, what is specific to webauth is that OAuth url must contain a temporary token generated by the connector and the resulting callback url, returned by BI will contain a new connection id which must be saved in the account.

### CODE REVIEW

File: packages/cozy-harvest-lib/src/services/budget-insight.spec.js [hunk 1]
```diff
@@ -60,7 +60,8 @@ const account = {
   _id: '1337',
   auth: {
     login: '80546578',
-    secret: 'secretsecret'
+    secret: 'secretsecret',
+    bankId: TEST_BANK_COZY_ID
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/src/services/budget-insight.spec.js:66]: This does not seem to be tested, should we check that it is correctly passed to createTemporaryToken ?

File: packages/cozy-harvest-lib/src/components/OAuthForm.jsx [hunk 2]
```diff
@@ -17,21 +19,39 @@ export class OAuthForm extends PureComponent {
...
   }
 
   componentDidMount() {
-    const { account } = this.props
+    const { account, konnector, flow, client } = this.props
     this.setState({ initialValues: account ? account.oauth : null })
+
+    const konnectorPolicy = findKonnectorPolicy(konnector)
+
+    if (konnectorPolicy.fetchMoreOAuthUrlParams) {
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/src/components/OAuthForm.jsx:35]: s/More/Extra everywhere is more correct IMHO.

File: packages/cozy-harvest-lib/src/components/OAuthForm.jsx [hunk 3]
```diff
@@ -17,21 +19,39 @@ export class OAuthForm extends PureComponent {
...
+      this.setState({ needMoreParams: true })
+      konnectorPolicy.fetchMoreOAuthUrlParams({
+        flow,
+        konnector,
+        client,
+        onSuccess: this.handleMoreParams
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/src/components/OAuthForm.jsx:41]: Is there a particular reason here to not use Promises ?

const extraParams = await konnectorPolicy.fetchExtraOAuthUrlParams({...})
this.setState({ extraParams })

File: packages/cozy-harvest-lib/src/helpers/oauth.js [hunk 4]
```diff
@@ -100,7 +100,8 @@ export const getOAuthUrl = ({
   oAuthStateKey,
   oAuthConf,
   nonce,
-  redirectSlug
+  redirectSlug,
+  moreParams
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/src/helpers/oauth.js:104]: This can be named extraParams. This could be documented in the docstring.

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx [hunk 5]
```diff
@@ -97,16 +97,35 @@ export class DumbTriggerManager extends Component {
    * OAuth Form success handler. OAuthForm retrieves an account id created by the
    * cozy stack
    * @param  {string}  accountId
+   * @param  {integer} bankId
    */
-  async handleOAuthAccountId(accountId) {
+  async handleOAuthAccountId(accountId, bankId) {
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/src/components/TriggerManager.jsx:102]: I am a bit surprised to see the banking notion leak here. I cannot see where this is called though. Would it be possible to have only generic oauth handling in TriggerManager and keep banking specific behaviors in the policy ?

File: packages/cozy-harvest-lib/src/services/budget-insight.js [hunk 6]
```diff
@@ -200,6 +200,51 @@ export const getBIConnectionId = account => {
   return get(account, 'data.auth.bi.connId')
 }
 
+export const handleWebauthAccount = async ({
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/src/services/budget-insight.js:203]: Can you add a docstring ?

File: packages/cozy-harvest-lib/src/services/budget-insight.js [hunk 7]
```diff
@@ -200,6 +200,51 @@ export const getBIConnectionId = account => {
...
+  return connectionId
+}
+
+/**
+ * Gets BI webauth connection id which is returned in the account by the stack
+ * via oauth callback url
+ *
+ * @param {io.cozy.accounts} The account content created by the stack
+ *
+ * @return {Integer} Connection Id
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/src/services/budget-insight.js:242]: It does not seem to be an Integer here https://github.com/cozy/cozy-libs/pull/1118/commits/9f71f8cdf51eee7fc4954b24296f444cfcf93e97#diff-8f55641483795f28fabc9c030803dd6aR355, can you change it to String ?

File: packages/cozy-harvest-lib/src/services/budget-insight.spec.js [hunk 8]
```diff
@@ -298,3 +300,78 @@ describe('createOrUpdateBIConnection', () => {
...
+    })
+    const flow = new ConnectionFlow(client, { konnector, account })
+    flow.handleFormSubmit = jest.fn()
+    flow.saveAccount = async account => account
+    const account = { oauth: { query: { id_connection: ['12'] } } }
+    const konnector = {}
+    const t = jest.fn()
+    await handleWebauthAccount({
+      account,
+      bankId: 12,
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/src/services/budget-insight.spec.js:360]: Can you use a different id that the connection_id, I find it a bit confusing.

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx [hunk 9]
```diff
@@ -97,16 +97,35 @@ export class DumbTriggerManager extends Component {
...
-      trigger: flow.trigger,
-      t: t
-    })
+    let oAuthAccount = await fetchAccount(client, accountId)
+
+    const konnectorPolicy = findKonnectorPolicy(konnector)
+
+    let foundWebauthConnection = false
+    if (konnectorPolicy.handleWebauthAccount) {
+      foundWebauthConnection = await konnectorPolicy.handleWebauthAccount({
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/src/components/TriggerManager.jsx:110]: Webauth is very specific to BI. I think TriggerManager should not know about it. Could we find a generic name for the konnector policy method ?

File: packages/cozy-harvest-lib/src/helpers/accounts.js [hunk 10]
```diff
@@ -192,7 +192,7 @@ export const resetState = account => ({
  * @param  {object} account   io.cozy.accounts document
  * @return {object}           io.cozy.accounts updated document
  */
-export const setSessionResetIfNecessary = (account, changedFields) => {
+export const setSessionResetIfNecessary = (account, changedFields = {}) => {
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/src/helpers/accounts.js:195]: Is this change necessary ? I do not see any added tests ?

File: packages/cozy-harvest-lib/src/services/budget-insight.spec.js (hunk 1 above)
Comment (doubleface) [reply packages/cozy-harvest-lib/src/services/budget-insight.spec.js:66]: CreateTemporaryToken has an assert for the presence of this attribute. And had to make this modification to allow the test to pass

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx (hunk 5 above)
Comment (doubleface) [reply packages/cozy-harvest-lib/src/components/TriggerManager.jsx:102]: Right, and this is not needed. I will keep this bankId in BI policy

File: packages/cozy-harvest-lib/src/services/budget-insight.js (hunk 7 above)
Comment (doubleface) [reply packages/cozy-harvest-lib/src/services/budget-insight.js:242]: But this is an integer. This is the result of getWebauthBIConnectionId function

File: packages/cozy-harvest-lib/src/components/OAuthForm.jsx (hunk 3 above)
Comment (doubleface) [reply packages/cozy-harvest-lib/src/components/OAuthForm.jsx]: Right, no good reason

**Review by ptbrowne — APPROVED** [review]
